shared/src/TwinCollectionArray.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// Represents a property array in a <see cref="TwinCollection"/>
+    /// </summary>
+    public class TwinCollectionArray : JArray
+    {
+        private readonly JObject _metadata;
+
+        internal TwinCollectionArray(JArray jArray, JObject metadata)
+            : base(jArray)
+        {
+            _metadata = metadata ?? throw new ArgumentNullException(nameof(metadata));
+        }
+
+        /// <summary>
+        /// Gets the value for the given property name
+        /// </summary>
+        /// <param name="propertyName">Property Name to lookup</param>
+        /// <returns>Property value, if present</returns>
+        public dynamic this[string propertyName]
+        {
+            get
+            {
+                if (propertyName == TwinCollection.MetadataName)
+                {
+                    return GetMetadata();
+                }
+
+                if (propertyName == TwinCollection.LastUpdatedName)
+                {
+                    return GetLastUpdated();
+                }
+
+                if (propertyName == TwinCollection.LastUpdatedVersionName)
+                {
+                    return GetLastUpdatedVersion();
+                }
+
+                throw new ArgumentException($"{nameof(TwinCollectionArray)} does not contain a definition for '{propertyName}'.");
+            }
+        }
+
+        /// <summary>
+        /// Gets the Metadata for this property
+        /// </summary>
+        /// <returns>Metadata instance representing the metadata for this property</returns>
+        public Metadata GetMetadata()
+        {
+            return new Metadata(GetLastUpdated(), GetLastUpdatedVersion());
+        }
+
+        /// <summary>
+        /// Gets the LastUpdated time for this property
+        /// </summary>
+        /// <returns>DateTime instance representing the LastUpdated time for this property</returns>
+        public DateTime GetLastUpdated()
+        {
+            return (DateTime)_metadata[TwinCollection.LastUpdatedName];
```

Comment:
I think the return value needs to change to `DateTime?` and the return statement to `return (DateTime?)_metadata?[TwinCollection.LastUpdatedName];`

Reply:
Would you add a couple of unit tests to validate that this methods down throw when initialized with null?